README.md
[![Clojars Project](https://img.shields.io/clojars/v/jmh-clojure.svg)](https://clojars.org/jmh-clojure)
[![](https://github.com/jgpc42/jmh-clojure/workflows/Test%20runner/badge.svg)][ci]

### Dependency and version information
<details>
  <summary>Click to show</summary>

[Leiningen][lein]

``` clojure
[jmh-clojure "0.4.1"]
```

[tools.deps][deps]

```clojure
{jmh-clojure/jmh-clojure {:mvn/version "0.4.1"}}
```

[Maven][maven]

``` xml
<dependency>
  <groupId>jmh-clojure</groupId>
  <artifactId>jmh-clojure</artifactId>
  <version>0.4.1</version>
</dependency>
```

JDK versions 8 to 18 and Clojure versions 1.7 to 1.11 are currently [tested against][ci].
</details>

### What is it?

This library provides a data-oriented API to [JMH][jmh], the Java Microbenchmark Harness.

JMH is developed by OpenJDK JVM experts and goes to great lengths to ensure accurate benchmarks. Benchmarking on the JVM is a complex beast and, by extension, JMH takes a bit of effort to learn and use properly. That being said, JMH is very robust and configurable. If you are new to JMH, I would recommend browsing the [sample][samples] code and javadocs before using this library.

If you a need simpler, less strenuous tool, I would suggest looking at the popular [criterium][criterium] library.

### Quick start

As a simple example, let's say we want to benchmark our fn that gets the value at an arbitrary indexed type. Of course, the built in `nth` already does this, but we can't extend `nth` to existing types like `java.nio.ByteBuffer`, etc.

```clojure
(ns demo.core)

(defprotocol ValueAt
  (value-at [x idx]))

(extend-protocol ValueAt
  clojure.lang.Indexed
  (value-at [i idx]
    (.nth i idx))
  CharSequence
  (value-at [s idx]
    (.charAt s idx))
  #_...)
```

Benchmarks are [usually](#alternate-ways-to-run) described in data and are fully separated from definitions. The reason for this is twofold. First, decoupling is generally good design practice. And second, it allows us to easily take advantage of JMH process isolation (forking) for reliability and accurracy. More on this later.

For repeatability, we'll place the following data in a `benchmarks.edn` resource file in our project. (Note that using a file is not a requirement, we could also specify the same data in Clojure. The `:fn` key values would need to be quoted in that case, however.)

```clojure
{:benchmarks
 [{:name :str, :fn demo.core/value-at, :args [:state/string, :state/index]}
  {:name :vec, :fn demo.core/value-at, :args [:state/vector, :state/index]}]

 :states
 {:index {:fn (partial * 0.5), :args [:param/count]} ;; mid-point
  :string {:fn demo.utils/make-str, :args [:param/count]}
  :vector {:fn demo.utils/make-vec, :args [:param/count]}}

 :params {:count 10}}
```

I have omitted showing the `demo.utils` namespace for brevity, it is defined [here][utils] if interested.

The above data should be fairly easy to understand. It is also a limited view of what can be specified. The [sample file][sample] provides a complete reference and explanation.

Now to run the benchmarks. We'll start a REPL in our project and evaluate the following. Note that we could instead use [lein-jmh][lein-jmh] or one of the other [supported tools](#tooling-support) to automate this entire process.

```clojure
(require '[jmh.core :as jmh]
         '[clojure.java.io :as io]
         '[clojure.edn :as edn])

(def bench-env
  (-> "benchmarks.edn" io/resource slurp edn/read-string))

(def bench-opts
  {:type :quick
   :params {:count [31 100000]}
   :profilers ["gc"]})

(jmh/run bench-env bench-opts)
;; => ({:name :str, :params {:count 31},     :score [1.44959801438209E8 "ops/s"], #_...}
;;     {:name :str, :params {:count 100000}, :score [1.45485370497829E8 "ops/s"]}
;;     {:name :vec, :params {:count 31},     :score [1.45550038851249E8 "ops/s"]}
;;     {:name :vec, :params {:count 100000}, :score [8.5783753539823E7 "ops/s"]})
```

> **Note:** due to the way jmh-clojure [works][works], the `*compile-path*` directory should exist and be on your classpath before benchmarking. This is automated by tools like Leiningen. For `tools.deps`, see [here][task-cp].

The `run` fn takes a benchmark environment and an optional map. We select the `:quick` type: an [alias][alias-doc] for some common options. We override our default `:count` parameter sequence to measure our fn against small and large inputs. We also enable the gc profiler.

Notice how we have four results: one for each combination of parameter and benchmark fn. For this example, we have omitted lots of additional result map data, including the [profiler][profilers] information.

Note that the above results were taken from multiple [runs][result], which is always a good practice when benchmarking.

#### Alternate ways to run

Benchmarking expressions or fns manually without the data specification is also supported. For example, the `run-expr` macro provides an interface similar to criterium, and allows benchmarking of code that only resides in memory (that you are updating in a REPL, for example), rather than on disk (loadable via `require`). However, this forgoes JMH process isolation. For more on why benchmarking this way on the JVM can be sub-optimal, see [here][extended].

### Tooling support

This library can be used directly in a [bare REPL][bare], as shown above, or standalone via [tools.deps][cli]. For a more robust experience, see the [`jmh-clojure-task`][task] project. This companion library provides some additional convenience features like sorting, table output, easy uberjar creation, and more. It can be easily integrated with tools like [Leiningen][lein-jmh].

### More information

As previously mentioned, please see the [sample file][sample] for the complete benchmark environment reference. For `run` options, see the [docs][run-doc]. Also, see the [wiki][wiki] for additional examples and topics.

The materials for a talk I gave at a [London Clojurians][london] online meetup are also available [here][talk]. A video capture of the event can also be viewed on [YouTube][video].

### Running the tests

```bash
lein test
```

Or, `lein test-all` for all supported Clojure versions.

### License

Copyright © 2017-2023 Justin Conklin

Distributed under the Eclipse Public License, the same as Clojure.



[alias-doc]:  https://jgpc42.github.io/jmh-clojure/doc/jmh.option.html#var-*type-aliases*
[bare]:       https://github.com/jgpc42/jmh-clojure/wiki/Bare-REPL
[ci]:         https://github.com/jgpc42/jmh-clojure/blob/master/.github/workflows/test.yml
[cli]:        https://github.com/jgpc42/jmh-clojure/wiki/Clojure-1.9-CLI
[criterium]:  https://github.com/hugoduncan/criterium
[deps]:       https://github.com/clojure/tools.deps.alpha
[extended]:   https://github.com/jgpc42/jmh-clojure/wiki/Extended
[jmh]:        https://github.com/openjdk/jmh
[lein]:       http://github.com/technomancy/leiningen
[lein-jmh]:   https://github.com/jgpc42/lein-jmh
[london]:     https://www.meetup.com/London-Clojurians/
[maven]:      http://maven.apache.org
[profilers]:  https://github.com/jgpc42/jmh-clojure/wiki/JMH-Profilers
[result]:     https://gist.github.com/jgpc42/4d8a828f8d0739748afa71035f2b2c9c#file-results-edn
[run-doc]:    https://jgpc42.github.io/jmh-clojure/doc/jmh.core.html#var-run
[sample]:     https://github.com/jgpc42/jmh-clojure/blob/master/resources/sample.jmh.edn
[samples]:    https://github.com/openjdk/jmh/tree/41548a7/jmh-samples/src/main/java/org/openjdk/jmh/samples
[talk]:       https://github.com/jgpc42/london-clojurians-jmh-talk-2020
[task]:       https://github.com/jgpc42/jmh-clojure-task
[task-cp]:    https://github.com/jgpc42/jmh-clojure-task#usage
[utils]:      https://gist.github.com/jgpc42/4d8a828f8d0739748afa71035f2b2c9c#file-utils-clj
[video]:      https://www.youtube.com/watch?v=_6qVfFkBdWI
[wiki]:       https://github.com/jgpc42/jmh-clojure/wiki
[works]:      https://github.com/jgpc42/jmh-clojure/wiki/How-It-Works


CHANGELOG.md
## 0.4.2-SNAPSHOT (20XX-XX-XX)

* Upgrade JMH to version `1.37`
* Explicit support for JDKs 15-21
* Explicit support for Clojure 1.11
* Improve error message when benchmark `:name` is invalid Java method identifier ([#10][issue10])

## 0.4.1 (2021-08-24)

* Upgrade JMH to version `1.32`

## 0.4.0 (2020-08-19)

* Automatically merge `:options` environment keys as `:type` aliases for convenience
* Support multiple `:type` aliases as seq
* Downgrade `insn` to `0.2.1` to avoid classpath collisions with (many popular) libraries that use older `ASM` versions
* Upgrade JMH to latest version `1.24`
* Rename `:profilers` result key to the more accurate `:secondary`
* Return `:statistics` data in all cases that JMH does
* Provide `:histogram` data in `:statistics`
* BUGFIX: warmup `:select`s should only be run during warmup
* BUGFIX: emit all selected benchmarks to single class to enable JMH `:group` functionality
* BUGFIX: improper single-width prim loads with `:instrument` option
* BUGFIX: option normalization should be done once at the end when all maps are merged

## 0.3.1 (2020-07-19)

* BUGFIX: compile Java code for advertized baseline JDK 8 compatibility

## 0.3.0 (2020-07-14)

* Upgrade JMH to latest version `1.23`
* Add a `deps.edn` file

## 0.2.1 (2017-12-29)

* Include new benchmark result key `:score-confidence`
* Include extra `:statistics` info for benchmark and profiler results when available
* Show original parameter names for benchmark result `:args`, not the gensyms

## 0.2.0 (2017-12-17)

* Support external profiler classes and profiler registry
* BUGFIX: `:java` should be subkey of `:jvm` and not at top-level of `:fork` option

## 0.1.7 (2017-11-09)

* Default `:fail-on-error` option to `true`
* Improve error messages for `:fn` expressions

## 0.1.6 (2017-10-20)

* BUGFIX: sometimes score units would be missing in result maps
* Better benchmark and state `:fn` validation

## 0.1.5 (2017-10-07)

* Undefined `:select` keywords to `jmh.core/run` use implicit selectors ([#2][issue2])
* Treat string `:params` as normal strings

## 0.1.4 (2017-10-06)

* BUGFIX: `:synchronize` should default to `true` when option `:iterations` is passed to `jmh.core/run`

## 0.1.3 (2017-10-02)

* BUGFIX: exception when passing global time options (e.g., `:timeout`) to `jmh.core/run`
* Use `:expr` not `"expr"` as the default name for benchmark `:fn` expressions

## 0.1.2 (2017-10-02)

* BUGFIX: [#1][issue1] NPE when no benchmarks are defined/selected

## 0.1.1 (2017-09-27)

* Compile java files with `-target 1.6` for older JVM compatibility



[issue1]:   https://github.com/jgpc42/jmh-clojure/issues/1
[issue2]:   https://github.com/jgpc42/jmh-clojure/issues/2
[issue10]:  https://github.com/jgpc42/jmh-clojure/issues/10


project.clj
(def dependencies
  (->> "deps.edn" slurp read-string
       :deps (mapv #(vector (% 0) (:mvn/version (% 1))))
       (into '[[org.clojure/clojure "1.8.0"]])))

(def javac-options
  (let [spec (-> (System/getProperty "java.specification.version")
                 Double/valueOf)]
    (if (= spec 1.8)
      []
      `["-target" "1.8"
        "-source" "1.8"
        ~@(when-let [root (System/getenv "JAVA_8_HOME")]
            [(str "-Xbootclasspath:"
                   (-> (java.io.File. root "jre/lib/rt.jar") .getCanonicalPath))])])))

(defproject jmh-clojure "0.4.2-SNAPSHOT"
  :description "Benchmarking with JMH, the Java Microbenchmark Harness, from Clojure."
  :url "https://github.com/jgpc42/jmh-clojure"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}

  :dependencies ~dependencies

  :min-lein-version "2.0.0"
  :jar-exclusions [#".+\.java$"]
  :java-source-paths ["java"]
  :javac-options ~javac-options
  :test-selectors {:unit (complement :integration)}

  :aliases {"test-all" ["do" "javac," "test,"
                        "with-profile" "+1.11" "test,"
                        "with-profile" "+1.10" "test,"
                        "with-profile" "+1.9" "test,"
                        "with-profile" "+1.7" "test"]}

  :profiles
  {:1.7 {:dependencies [[org.clojure/clojure "1.7.0"]]}
   :1.9 {:dependencies [[org.clojure/clojure "1.9.0"]]}
   :1.10 {:dependencies [[org.clojure/clojure "1.10.0"]]}
   :1.11 {:dependencies [[org.clojure/clojure "1.11.0"]]}
   :repl {:source-paths ["dev"]}})
